src/compressed_tensors/transform/apply.py

```diff
@@ -34,3 +38,41 @@ def apply_transform_config(model: torch.nn.Module, config: TransformConfig):
 
     # attach config to model for compression/serialization
     setattr(model, TRANSFORM_CONFIG_NAME, config)
+
+    # ensure that tied weight transforms can be serialized without aliases
+    # In the future, this could be done by transformers or model compressor
+    # which would make this more robust to changing dispatches after transforms
+    _tie_offloaded_tensors(model)
+
+
+def _tie_offloaded_tensors(model: torch.nn.Module):
+    """
+    When accelerate replaces tensors with meta tensors during offloading, the meta
+    tensors may not be identical, even if the offloaded values are identical.
+
+    However, transformers can only serialize correctly if meta tensors are identical
+    (see transformers#39263).
+
+    This function collects all meta tensors which have shared offloaded values and sets
+    those tensors to be identical so that they can be removed during serialization
+
+    :param model: model potentially containing offloaded meta tensors to fix
+    """
+
+    # map from offloaded tensor pointers to module-key locations
+    offloaded_ptrs: dict[int, List[Tuple[torch.nn.Module, str]]] = defaultdict(list)
+    for module in model.modules():
+        if has_offloaded_params(module):
+            for key, _ in module.named_parameters(recurse=False):
+                param = module._hf_hook.weights_map[key]
+                offloaded_ptrs[id(param)].append((module, key))
+
+    # ensure that if a location shares an offloaded tensor pointers, that the
+    # meta tensor is also identical (assigned to the first element of the set)
+    for shared_keys in offloaded_ptrs.values():
+        assert len(shared_keys) >= 1
+        first_tensor = getattr(shared_keys[0][0], shared_keys[0][1])
+        assert first_tensor.device.type == "meta"
+        for module, key in shared_keys:
+            assert getattr(module, key).device.type == "meta"
+            setattr(module, key, first_tensor)
```

Comment:
This could be simplified by having `offloaded_ptrs: dict[int, Tensor]` store the the meta tensor of the first occurrence of the id. Then you just loop through the modules/params. If it's a new param, store the meta_tensor else update the meta_tensor.

Note: you might have to use iterate through `list(module.named_parameters(recurse=False))` so that it doesn't get upset about you changing the values while iterating. 

Also fine as is, but I think this achieves the same thing and is a bit simpler.

Reply:
You would be right, except for that there is no way to "update" a meta tensor in place. This would involve changing a tensor's device inplace, which is not a valid operation.

You can only overwrite the tensor at that location, and `(module, key)` is a unique descriptor of that location.